File: codes/javascript/polymorphism/es6/subtype.js
#!/usr/bin/env node

class Person {
    talk() {
        console.log("I am a person");
    }
}

class Student extends Person {}

class Professor extends Person {}

class Robot {
    talk() {
        console.log("I am a robot");
    }
}

function announce(msg, talker) {
    process.stdout.write(msg + ": ");
    if (typeof talker.talk == "function") talker.talk();
    else console.log("Sorry! I do not talk!");
}

let pe = new Person(),
    st = new Student(),
    pr = new Professor();

let ro = new Robot();
let cat = {
    talk: function(){
        console.log("Meow!");
    }
};

announce("Person", pe);
announce("Student", st);
announce("Professor", pr);
announce("Robot", ro);
announce("a cat", cat);
announce("a number", 25);

console.log("A table with different types");
let elements = [pe, st, pr, ro, cat, 25];
for (let i = 0; i < elements.length; i++){
    try {
        elements[i].talk();
    }
    catch(error){
        console.log("The element n° ", i, " does not talk()");
    }
}

File: codes/lua/polymorphism/subtype.lua
-- Person class
local Person = {}
Person.__index = Person

function Person:new()    -- The constructor
    local instance = {}
    setmetatable(instance, self) --self, here, will refer to the calling class
    return instance
end

function Person:talk()
    print("I am a person")
end

-- Student class
local Student = {}
Student.__index = Student
setmetatable(Student, Person) -- extends

-- Professor class
local Professor = {}
Professor.__index = Professor
setmetatable(Professor, Person) -- extends

function announce(msg, talker)
    io.write(msg .. ": ")
    if type(talker) == "table" and type(talker.talk) == "function" then
        talker:talk()
    else
        print("Sorry! I do not talk!")
    end
end

local pe = Person:new()
local st = Student:new()
local pr = Professor:new()

local cat = {
    talk = function(); print("Meow!"); end
}

announce("Person", pe)
announce("Student", st)
announce("Professor", pr)
announce("a cat", cat)
announce("a number", 25)

print("A table with different types")
local elements = {pe, st, pr, cat, 25}
for i =1, #elements do
    success, res = pcall(function() elements[i].talk(); end)
    if not success then
        print("The element n°" .. i .. " does not talk()")
    end
end
